### PR TITLE
make power state led configurable on PlugS

### PR DESCRIPTION
**Pull request by timoschilling**
Make the power state led of the PlugS configurable, closes https://github.com/mongoose-os-apps/shelly-homekit/issues/342.

Current testet on a PlugS, but maybe we can expand that pattern to even more shelly types.

### CODE REVIEW

**Comment by rojer** [comment]
thanks Timo! this looks good but we need this option to not be displayed at all if there is no LED to speak of (i.e. if led_out_ is null).

**Comment by timoschilling** [comment]
Yes you are right, I noticed it by my self, I'm working on a fix.
I switched `power_state_led` from an bool to an int:
- -1 means UI didn't shown
- 0 means LED off
- 1 means LED on

**Comment by timoschilling** [comment]
updated

File: mos.yml [hunk 1]
```diff
@@ -48,6 +48,7 @@ config_schema:
   - ["sw.initial_state", "i", 3, {title: "Initial state on power-on: 0 - off, 1 - on, 2 - restore last state, 3 - matches input if in toggle mode, otherwise off"}]
   - ["sw.auto_off", "b", false, {title: "Whether the switch should automatically turn OFF after turning ON"}]
   - ["sw.auto_off_delay", "d", 0, {title: "Delay for automatically turning OFF, in seconds"}]
+  - ["sw.power_state_led", "i", -1, {title: "Power state LED enabled"}]
```
Comment (rojer) [inline mos.yml:51]: explain that `-1` means no power led.

**Review by rojer — COMMENTED** [review]
pretty good, just a minor comment. plus resolve the index.html.gz conflict and we're good to go.

File: mos.yml [hunk 2]
```diff
@@ -48,7 +48,7 @@ config_schema:
   - ["sw.initial_state", "i", 3, {title: "Initial state on power-on: 0 - off, 1 - on, 2 - restore last state, 3 - matches input if in toggle mode, otherwise off"}]
   - ["sw.auto_off", "b", false, {title: "Whether the switch should automatically turn OFF after turning ON"}]
   - ["sw.auto_off_delay", "d", 0, {title: "Delay for automatically turning OFF, in seconds"}]
-  - ["sw.power_state_led", "i", -1, {title: "Power state LED enabled"}]
+  - ["sw.power_state_led", "i", -1, {title: "Power state LED: -1 - unsupported by device, 0 - off, 1 - on"}]
```
Comment (rojer) [inline mos.yml:51]: let's call this `sw.state_led_en`. -1 - unsupported, 0 - disabled, 1 - enabled

please update other places accordingly

File: fs_src/index.html [hunk 3]
```diff
@@ -282,6 +282,13 @@ <h1 id="head">Switch</h1>
         <input type="text" id="auto_off_delay" placeholder="D:HH:MM:SS.sss" required
                pattern="[0-9]+:(0[0-9]|1[0-9]|2[0-3]):[0-5][0-9]:[0-5][0-9]\.[0-9]{3}">
       </div>
+      <div class="form-control" id="power_state_led_container">
+        <label for="power_state_led">Power state led:</label>
```
Comment (rojer) [inline fs_src/index.html:286]: State LED

**Review by rojer — COMMENTED** [review]
almost there

File: mos.yml (hunk 2 above)
Comment (timoschilling) [reply mos.yml:51]: I can do, I only want to say that there are different status LEDs in the Shelly's. Power in the PlugS, Connect/Paired in the PlugS, Wifi in the Dimmer/RGBW. So I think we should be more precise with the naming.

Comment (rojer) [reply mos.yml:51]: but this one will be under specific switch section, i think it's clear. we can have a setting under system settings for system led indication.

**Comment by timoschilling** [comment]
@rojer it’s updated

File: src/shelly_switch.cpp [hunk 4]
```diff
@@ -138,6 +139,10 @@ Status ShellySwitch::SetConfig(const std::string &config_json,
   cfg_->initial_state = cfg.initial_state;
   cfg_->auto_off = cfg.auto_off;
   cfg_->auto_off_delay = cfg.auto_off_delay;
+  if (cfg_->state_led_en != cfg.state_led_en) {
```
Comment (rojer) [inline src/shelly_switch.cpp:142]: i think we should disallow setting to any value if it's currently -1.

Comment (timoschilling) [reply src/shelly_switch.cpp:142]: Like this?
```suggestion
  if (cfg_->state_led_en != -1 && cfg_->state_led_en != cfg.state_led_en) {
```

Comment (rojer) [reply src/shelly_switch.cpp:142]: even earlier, during validation, we should check for valid new values - 0, 1 and that previous not -1, and send an error

Comment (timoschilling) [reply src/shelly_switch.cpp:142]: would you like add append an commit with a change according to your ideas

Comment (rojer) [reply src/shelly_switch.cpp:142]: i don't think i can push directly to this branch, but here's what i meant:

```
  if ((cfg_->state_led_en == -1 && cfg.state_led_en != -1) ||
      (cfg_->state_led_en != -1 && cfg.state_led_en != 0 &&
       cfg.state_led_en != 1)) {
    return mgos::Errorf(STATUS_INVALID_ARGUMENT, "invalid %s", "state_led_en");
  }
```

File: src/shelly_switch.cpp [hunk 5]
```diff
@@ -115,6 +116,9 @@ Status ShellySwitch::SetConfig(const std::string &config_json,
   if (cfg.initial_state < 0 || cfg.initial_state > (int) InitialState::kMax) {
     return mgos::Errorf(STATUS_INVALID_ARGUMENT, "invalid %s", "initial_state");
   }
+  if (cfg.state_led_en < 0 || cfg.state_led_en > 1) {
+    return mgos::Errorf(STATUS_INVALID_ARGUMENT, "invalid %s", "state_led_en");
+  }
```
Comment (timoschilling) [inline src/shelly_switch.cpp:121]: @rojer like this?

Comment (rojer) [reply src/shelly_switch.cpp:121]: not quite, see my comment above

File: src/shelly_switch.cpp (hunk 4 above)
Comment (timoschilling) [reply src/shelly_switch.cpp:147]: updated my branch

**Comment by rojer** [comment]
merged, thank you!

**Comment by timoschilling** [comment]
🎉 nice to be part of this great project!!!

**Comment by eSportler89** [comment]
Good morning, with Version _2.9.1_, Build _20210518-010122/2.9.1-gb7dcbef-master_ I got the problem of a blinking blue LED. Changing the Powerstate LED works fine, but is there a way to disable the wifi status light?
Normally it would be solid when connected and blinking when it's in AP mode.

Thanks!

**Comment by timoschilling** [comment]
The blue LED is off when connected to a WiFi and connected to HomeKit

**Comment by eSportler89** [comment]
Argh, damn!
Configured it, connected to my wifi but didn't add it to my Homekit.... I only tested it within the webinterface.
Thank you! :)

**Pull request merged**